### PR TITLE
[browser] Marshalling support float[], Span<float> and ArraySegment<float>

### LINKED ISSUE #97380: [browser][JSImport] Support for marshaling float memory views
Support for `Span<Single>` and `ArraySegment<Single>` in [JSImport]/[JSExport]

As requested in https://github.com/dotnet/runtime/discussions/97268

### LINKED ISSUE #123706: [API Proposal]: JavaScript marshalling support float[], Span<float> and ArraySegment<float>
### Background and motivation

Ability to access floats in the WASM linear memory from in WebGPU contexts, without requiring data copies.
Performance of rendering and also neural networks (AI) depends on no-copy interop.
In those scenarios buffers of `float[]` are in many cases preferred over `double[]` buffers because of memory bandwidth and size of RAM on the GPU.

See https://github.com/dotnet/runtime/discussions/97268 and https://github.com/dotnet/runtime/issues/97380.

We already have the same API for copy-by-value JS interop of `double[]` and no-copy interop of , `Span<double>`, `ArraySegment<double>`

### API Proposal

```diff
namespace System.Runtime.InteropServices.JavaScript;

[Versioning.SupportedOSPlatformAttribute("browser")]
[CLSCompliant(false)]
[System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
public struct JSMarshalerArgument
{
    ....
+   public void ToManaged(out float[]? value) { throw null; }
    public void ToManaged(out double[]? value) { throw null; }
+   public void ToJS(float[]? value) { throw null; }
    public void ToJS(double[]? value) { throw null; }
+   public void ToManaged(out Span<float> value) { throw null; }
    public void ToManaged(out Span<double> value) { throw null; }
+   public void ToJS(Span<float> value) { throw null; }
    public void ToJS(Span<double> value) { throw null; }
+   public void ToManaged(out ArraySegment<float> value) { throw null; }
    public void ToManaged(out ArraySegment<double> value) { throw null; }
+   public void ToJS(ArraySegment<float> value) { throw null; }
    public void ToJS(ArraySegment<double> value) { throw null; }
}
```

### API Usage

This is **API for code generated by Roslyn** analyzer when users use `[JSImport]` or `[JSExport]`.
It's not human facing API, but it needs to be public API because Roslyn generated code is consuming only visible APIs.

The generated code looks similar to this, just with `float` data type.
https://github.com/dotnet/runtime/blob/cf05e728871742f5c5483db2f3efea537ff17328/src/libraries/System.Runtime.InteropServices.JavaScript/tests/JSImportGenerator.UnitTest/Compiles.cs#L210-L227

### User facing API

**This doesn't change** because this is all user code.

```csharp
[JSImport("echo1", "JavaScriptTestHelper")]
static partial float[]? Echo1(float[]? value);

[JSExport]
internal static ArraySegment<float> Echo2(ArraySegment<float> value)
{
    return value;
}

[JSImport("echo3", "JavaScriptTestHelper")]
static partial Span<float> Echo3(Span<float> value);

[JSExport]
static Span<float> EchoSpanSingle(Span<float> value)
{
    return value;
}
```

There is draft implementation https://github.com/dotnet/runtime/pull/123642

### Alternative Designs

_No response_

### Risks

No known risks.

### PR DESCRIPTION
**Pull request by ArcadeMode**
Fixes https://github.com/dotnet/runtime/issues/97380.
Fixes https://github.com/dotnet/runtime/issues/123706

Implementation is functional for CoreCLR and Mono. Added import and export tests to demonstrate.

How do I go about updating the docs? (and which version might see these changes for that matter, net10/11?)

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Marshaling/JSMarshalerArgument.Object.cs [hunk 1]
```diff
@@ -287,6 +287,11 @@ public void ToJS(object? value)
                 int[] val = (int[])value;
                 ToJS(val);
             }
+            else if (typeof(float[]) == type)
+            {
+                float[] val = (float[])value;
+                ToJS(val);
+            }
```
[the remark this answers is not in the document]
Comment (ArcadeMode) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Marshaling/JSMarshalerArgument.Object.cs:304]: extended ToManaged instead to align implementations

**Comment by ArcadeMode** [comment]
> > is it okay to ask for API review in PR like this?
> 
> The process is described here https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md
> 
> You need to create dedicated issue according to the issue template for [API suggestion](https://github.com/dotnet/runtime/issues/new/choose).
> 
> You can also look at successfully [approved APIs](https://github.com/dotnet/runtime/issues?q=is%3Aissue%20label%3Aapi-approved)

Thanks for the pointers. I have created https://github.com/dotnet/runtime/issues/123706 with the proposal

**Comment by pavelsavara** [comment]
test for JS Array values of wrong type or JS number (double)  above float.max would be great

**Comment by ArcadeMode** [comment]
> test for JS Array values of wrong type or JS number (double) above float.max would be great

I have created https://github.com/dotnet/runtime/issues/123731 to resolve issues with type assertions for array elements. Here I would only end up adding demonstration that the float would overflow (as demonstrated for int32). I hope I can address that issue separately.

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/System/Runtime/InteropServices/JavaScript/JavaScriptTestHelper.cs [hunk 2]
```diff
@@ -223,6 +223,14 @@ internal static partial void Relaxed(string a1, Exception ex,
         [return: JSMarshalAs<JSType.Number>]
         internal static partial double? store_DoubleArray([JSMarshalAs<JSType.Array<JSType.Number>>] double[]? value, [JSMarshalAs<JSType.Number>] int index);
 
+        [JSImport("echo1", "JavaScriptTestHelper")]
```
Comment (pavelsavara) [inline src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/System/Runtime/InteropServices/JavaScript/JavaScriptTestHelper.cs:231]: Could you please add few signatures without `JSMarshalAs` annotations. They should work fine too.

Comment (ArcadeMode) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/System/Runtime/InteropServices/JavaScript/JavaScriptTestHelper.cs:231]: I've added tests using imports without the annotations, done for both float[] and double[] as it was not covered for other array types either. span/segment excluded given they're not supported without annotations.

**Comment by ArcadeMode** [comment]
@pavelsavara just a heads up that I've adressed your review comment

**Comment by stephentoub** [comment]
## Code Review: PR #123642

### Holistic Assessment

**Motivation:** This PR adds marshalling support for `float[]`, `Span<float>`, and `ArraySegment<float>` in the JS interop layer, addressing issues #97380 and #123706. This fills a gap where `double` variants were supported but `float` (Single) was not.

**Approach:** The implementation follows the established patterns for other numeric types (byte, int32, double), consistently extending:
- The JSImport generator
- C# marshaling code in `JSMarshalerArgument.Single.cs`
- TypeScript marshaling code for both Mono browser runtime and the native interop library
- Corresponding test coverage

**Net Positive:** ✅ Yes - This is a clean, well-structured feature addition that follows existing conventions.

---

### Detailed Findings

#### ✅ Correctness
- The C# marshaling logic in `JSMarshalerArgument.Single.cs` correctly uses `sizeof(float)` (4 bytes) for offset calculations and memory operations.
- TypeScript code uses `Float32Array` and `localHeapViewF32()` appropriately for single-precision floats.
- The `fixupPointer(buffer_ptr, 2)` alignment (4-byte) is correct for `float`/`Single`.
- The `MemoryViewType.Single = 3` enum value is added consistently across all marshaled-types definitions.

#### ✅ Consistency with Existing Patterns
- The implementation mirrors the existing `double[]`, `Span<double>`, and `ArraySegment<double>` support.
- Generator factory additions follow the exact same pattern as `Double` entries.
- Test methods follow existing naming and structure conventions.

#### ✅ API Surface
- New public API methods added to ref assembly:
  - `ToManaged(out float[]? value)` / `ToJS(float[]? value)`
  - `ToManaged(out Span<float> value)` / `ToJS(Span<float> value)`
  - `ToManaged(out ArraySegment<float> value)` / `ToJS(ArraySegment<float> value)`
- These are consistent with existing overloads for other types.

#### ✅ Test Coverage
- Good test coverage including:
  - Array round-trip tests with explicit attributes and without
  - Span and ArraySegment tests
  - JSExport tests
  - Edge cases: `float.MaxValue`, `float.MinValue`, `float.Pi`, `float.NaN`, infinities

#### 💡 Suggestions (Non-blocking)

1. **Minor:** In `JSMarshalerType.cs` lines 488-489, an unrelated blank line was removed. This is a minor cosmetic cleanup but unrelated to the feature.

2. **Documentation:** The PR author asks about updating docs - typically this would be in the [dotnet/docs](https://github.com/dotnet/docs) repository, targeting the version where this ships (likely .NET 10).

#### ✅ Code Style
- The refactoring of `array_element_size` from nested ternary to if-statements improves readability.
- The refactoring of `_unsafe_create_view` similarly improves readability.

---

### Summary

**Verdict: ✅ Approve**

This PR is a clean, well-implemented feature that follows established patterns. The marshalling support for `float[]`, `Span<float>`, and `ArraySegment<float>` is correctly implemented across all layers (C# generator, C# runtime, TypeScript runtime). Test coverage is comprehensive. No correctness issues identified.

**Review by pavelsavara — APPROVED** [review]

**Review by maraf — APPROVED** [review]

**Comment by pavelsavara** [comment]
@danroth27 this is release-notes-worthy community contribution 💙 for Net11 preview 3 with doc update here https://github.com/dotnet/AspNetCore.Docs/pull/36886/changes
